File: apps/main/src/lib/api/wallet.ts
import { env } from "../env";
import { formatCurrency } from "@fe-el-seka/shared";
import type { Locale } from "@fe-el-seka/shared";

const base = env.apiUrl;

function authHeaders(token: string) {
  return { Authorization: `Bearer ${token}` };
}

function jsonAuthHeaders(token: string) {
  return { Authorization: `Bearer ${token}`, "Content-Type": "application/json" };
}

export interface LedgerEntry {
  id: string;
  type:
    | "COMMISSION_DEBIT"
    | "ADMIN_CREDIT"
    | "ADMIN_DEBIT"
    | "SPONSORED_RIDE_CREDIT"
    | "SPONSORED_RIDE_REVERSAL"
    | "WITHDRAWAL_DEBIT"
    | "CASH_BACK_CREDIT"
    | "CASH_BACK_REVERSAL"
    | "POINTS_DISCOUNT_REIMBURSEMENT"
    | "CASH_BACK_REDEEMED";
  amount_egp: string;
  ride_id: string | null;
  booking_id: string | null;
  fuel_cost_egp_snapshot: string | null;
  note: string | null;
  created_at: string;
}

export interface WalletResponse {
  balance_egp: string;
  reserved_egp: string;
  available_egp: string;
  sponsored_earnings_egp: string;
  cash_back_points_egp: string;
  entries: LedgerEntry[];
  pagination: {
    page: number;
    per_page: number;
    total_entries: number;
    total_pages: number;
  };
}

export interface CashBackRedeemResponse {
  id: string;
  amount_egp: string;
  cash_back_points_egp: string;
  sponsored_earnings_egp: string;
  created_at: string;
}

export async function redeemCashBackPoints(token: string, amountEgp: string): Promise<CashBackRedeemResponse> {
  const res = await fetch(`${base}/api/v1/drivers/me/wallet/cash-back/redeem`, {
    method: "POST",
    headers: jsonAuthHeaders(token),
    body: JSON.stringify({ amount_egp: amountEgp }),
  });
  const json = await res.json();
  if (!res.ok) throw json;
  return json;
}

export async function getWallet(token: string, page = 1): Promise<WalletResponse> {
  const res = await fetch(`${base}/api/v1/drivers/me/wallet?page=${page}`, {
    headers: authHeaders(token),
  });
  const json = await res.json();
  if (!res.ok) throw json;
  return json;
}

export function formatEgp(amount: string | number, locale: Locale = "en"): string {
  return formatCurrency(typeof amount === "string" ? parseFloat(amount) : amount, locale);
}

File: services/api/app/models/wallet.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Optional
from uuid import UUID

from pydantic import BaseModel, field_validator


class LedgerEntryType(str, Enum):
    COMMISSION_DEBIT = "COMMISSION_DEBIT"
    ADMIN_CREDIT = "ADMIN_CREDIT"
    ADMIN_DEBIT = "ADMIN_DEBIT"
    SPONSORED_RIDE_CREDIT = "SPONSORED_RIDE_CREDIT"
    SPONSORED_RIDE_REVERSAL = "SPONSORED_RIDE_REVERSAL"
    WITHDRAWAL_DEBIT = "WITHDRAWAL_DEBIT"
    CASH_BACK_CREDIT = "CASH_BACK_CREDIT"
    CASH_BACK_REVERSAL = "CASH_BACK_REVERSAL"
    POINTS_DISCOUNT_REIMBURSEMENT = "POINTS_DISCOUNT_REIMBURSEMENT"
    CASH_BACK_REDEEMED = "CASH_BACK_REDEEMED"


# ── Response schemas ─────────────────────────────────────────────────────────

class LedgerEntryResponse(BaseModel):
    id: UUID
    type: LedgerEntryType
    amount_egp: str
    ride_id: Optional[UUID]
    booking_id: Optional[UUID]
    fuel_cost_egp_snapshot: Optional[str]
    created_by: Optional[UUID]
    note: Optional[str]
    created_at: datetime


class PaginationMeta(BaseModel):
    page: int
    per_page: int
    total_entries: int
    total_pages: int


class WalletPageResponse(BaseModel):
    balance_egp: str
    reserved_egp: str
    available_egp: str
    sponsored_earnings_egp: str
    cash_back_points_egp: str
    entries: list[LedgerEntryResponse]
    pagination: PaginationMeta


# ── Cash Back redeem schemas ─────────────────────────────────────────────────

class CashBackRedeemRequest(BaseModel):
    amount_egp: Decimal

    @field_validator("amount_egp")
    @classmethod
    def amount_must_be_positive(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("amount_egp must be greater than 0.00 EGP")
        return v


class CashBackRedeemResponse(BaseModel):
    id: UUID
    amount_egp: str
    cash_back_points_egp: str
    sponsored_earnings_egp: str
    created_at: datetime


# ── Admin request / response schemas ─────────────────────────────────────────

class TopUpRequest(BaseModel):
    amount_egp: Decimal
    note: Optional[str] = None

    @field_validator("amount_egp")
    @classmethod
    def amount_must_be_positive(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("amount_egp must be greater than 0.00 EGP")
        return v


class TopUpResponse(BaseModel):
    wallet_id: UUID
    driver_id: UUID
    new_balance_egp: str
    ledger_entry_id: UUID
    amount_credited_egp: str
    created_at: datetime


class AdjustRequest(BaseModel):
    amount_egp: Decimal
    note: Optional[str] = None

    @field_validator("amount_egp")
    @classmethod
    def amount_must_be_positive(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("amount_egp must be greater than 0.00 EGP")
        return v


class AdjustResponse(BaseModel):
    wallet_id: UUID
    driver_id: UUID
    new_balance_egp: str
    new_available_egp: str
    ledger_entry_id: UUID
    amount_debited_egp: str
    created_at: datetime
